src/Measurement_Observation/haskell_model.rst

```diff
@@ -0,0 +1,129 @@
+type of the heap object. Complex forms are lowered to simpler ones (e.g. unboxed
+sums are lowered into unboxed tuples); this is called unarisation.
+
+Additionally every heap allocation is made explicit with a let-binding. For
+example instead of ``foo (MkBar x y)`` you have ``let b = MkBar x y in foo b``.
+This is called `A-normal form<https://en.wikipedia.org/wiki/A-normal_form>`_.
+During code generation from STG we know that all functions are applied to
+"simple" arguments only (variables, constants, etc.).
+
+Following the approach pioneered with "super-combinators", every top-level STG
+binding is then compiled into imperative code. The idea is that executing this
+imperative code has the same result as interpreting the functional code. For
+example, ``let b = MkBar x y in foo b`` is basically compiled to:
+
+.. code-block:: C
+
+  b <- allocate heap object for `MkBar x y`
+  evaluate foo
+  apply `foo` to `b`
+
+Different GHC backends use different imperative representations. The interpreter
+uses ByteCode, the JavaScript backend uses some JavaScript-like form, all other
+backends use Cmm.
+
+Cmm: an imperative language that looks like LLVM IR. It supports expressions,
+statements, and it abstracts over machine primops, machine registers, stack
+usage, and calling conventions. A pass performs register assignment and stack
+layout for the target architecture. Then assembly code (textual) is generated
+for the target architecture and an external assembler program generates machine
+code (binary) for it. Finally an external linker is used to transform the
+resulting code objects into a single executable or library.
+
+Executables are linked against a runtime system (RTS) that provides primitives
+to manage:
+- memory: allocation, garbage collection...
+- scheduling: thread scheduling, blocking queues...
+- IOs: primitives to interact with the operating system
+- dynamic code loading: loading and unloading code objects at runtime.
+
+The RTS itself comes in different flavors: e.g. using multiple OS threads to execute
+Haskell code or not.
+
+All these compilation stages and the RTS provide knobs to tweak the generated
+code and the behavior of the runtime system. In particular, some probes can be
+optionally inserted in the generated code at various stages to produce different
+profiling information at runtime.
+
+
+Execution model overview
+------------------------
+
+The heap of a Haskell program contain objects that reflect its current execution
+state: suspended computations (thunks), partially applied functions (PAP),
+values (datacon and their payload), threads (TSO) and their stack, etc.
+When the computer executes the compiled code for a top-level STG binding
+(starting from ``main``), it asks for more objects to be allocated in the heap,
+for some thunks to be reduced to another heap object, or for existing objects to
+be used as arguments for function applications.
```

Comment:
```suggestion
The heap of a Haskell program contain objects that reflect its current execution
state: suspended computations (:term:`thunk`), partially applied functions (:term:`PAP`),
values (data constructors and their payload), threads (:term:`TSO`) and their stack, etc.
When the computer executes the compiled code for a :term:`top-level` STG binding
(starting from ``main``), it asks for more objects to be allocated in the heap,
for some thunks to be reduced to another heap object, or for existing objects to
be used as arguments for function applications.
```

Reply:
The last few sentences also makes it sound as if the computer is executing STG and not assembly. I think its probably ok though.